pkg/controller/datasource-controller.go

```diff
@@ -126,19 +128,72 @@ func NewDataSourceController(mgr manager.Manager, log logr.Logger, installerLabe
 	if err != nil {
 		return nil, err
 	}
-	if err := addDataSourceControllerWatches(mgr, DataSourceController, log); err != nil {
+	if err := addDataSourceControllerWatches(mgr, DataSourceController); err != nil {
 		return nil, err
 	}
 	log.Info("Initialized DataSource controller")
 	return DataSourceController, nil
 }
 
-func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller, log logr.Logger) error {
+func addDataSourceControllerWatches(mgr manager.Manager, c controller.Controller) error {
 	if err := cdiv1.AddToScheme(mgr.GetScheme()); err != nil {
 		return err
 	}
-	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{}); err != nil {
+	if err := c.Watch(&source.Kind{Type: &cdiv1.DataSource{}}, &handler.EnqueueRequestForObject{},
+		predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool { return true },
+			DeleteFunc: func(e event.DeleteEvent) bool { return true },
+			UpdateFunc: func(e event.UpdateEvent) bool { return !sameDataSourcePvc(e.ObjectOld, e.ObjectNew) },
+		},
+	); err != nil {
+		return err
+	}
+
+	const dataSourcePvcField = "spec.source.pvc"
+
+	getKey := func(namespace, name string) string {
+		return namespace + "." + name
+	}
+
+	if err := mgr.GetFieldIndexer().IndexField(context.TODO(), &cdiv1.DataSource{}, dataSourcePvcField, func(obj client.Object) []string {
+		if pvc := obj.(*cdiv1.DataSource).Spec.Source.PVC; pvc != nil {
+			return []string{getKey(pvc.Namespace, pvc.Name)}
```

Comment:
It looks like you can just give the name and the library will add namespace prefix:  https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/informer_cache.go#L180-L217

Reply:
pvc.Namespace can be != DataSource.Namespace, and there can be same pvc names on different ns, so NO way the library can add the pvc ns prefix.

Reply:
We cannot use the `DataSource.namespace` for the pvc, since we don't know the `DataSource.namespace` when we call `getKey()` for creating `client.MatchingFields` for the `List()`. However, since we don't have  `dataSourceValidatingWebhook`, when `DataSource.spec.source.pvc` is updated and its namespace is emty we can update it to the DataSource namespace. Sounds ok?

Reply:
I don't understand.  How is the datasource namespace relevant to the list call?  Take the following:

```yaml
apiversion: cdi.kubevirt.io/v1beta1
kind: DataSource
metadata:
  namespace: ns1
  name: ds1
  spec:
    source:
      pvc:
        name: dv1
  ...
---
apiversion: cdi.kubevirt.io/v1beta1
kind: DataVolume:
metadata:
  namespace: ns1
  name: dv1
 ...
```


Given the current code, the index is:

`/dv1` -> [`ns1/ds1`]

and

`mapToDataSource` will lookup `ns1/dv1` and should come up empty, no?

Reply:
Right, my mistake. Fixing.